cocos/renderer/CCGLProgram.cpp

```diff
@@ -662,17 +662,23 @@ bool GLProgram::updateUniformLocation(GLint location, const GLvoid* data, unsign
     {
         GLvoid* value = malloc(bytes);
         memcpy(value, data, bytes );
-        _hashForUniforms.insert(std::make_pair(location, value));
+        _hashForUniforms.insert(std::make_pair(location, std::make_pair(value, bytes)));
     }
     else
     {
-        if (memcmp(element->second, data, bytes) == 0)
+        if (memcmp(element->second.first, data, bytes) == 0)
         {
             updated = false;
         }
         else
         {
-            memcpy(element->second, data, bytes);
+            if (element->second.second != bytes){
```

Comment:
when new length greater than old length, can realloc the buffer. If smaller, copy directly.